src/main/scala/rx/lang/scala/observables/AsyncOnSubscribe.scala

```diff
@@ -0,0 +1,120 @@
+package rx.lang.scala.observables
+
+import rx.annotations.Experimental
+import rx.lang.scala.{Notification, Observable, Observer}
+
+/**
+  * An utility class to create observables that start acting when subscribed to and respond correctly
+  * to back pressure requests from subscribers.
+  */
+object AsyncOnSubscribe {
+  /**
+    * Generates an `Observable` that synchronously calls the provided `next` function
+    * to generate data to downstream subscribers.
+    *
+    * @tparam T the type of the generated values
+    * @tparam S the type of the associated state with each Subscriber
+    * @param generator generates the initial state value
+    * @param next produces data to the downstream subscriber
+    * @param onUnsubscribe clean up behavior
+    * @return An Observable that emits data downstream in a protocol compatible with back-pressure.
+    * @see rx.observables.AsyncOnSubscribe.createStateful
+    */
+  @Experimental
+  def createStateful[S, T](generator: () => S)(next: (S, Long, Observer[Observable[_ <: T]]) => S, onUnsubscribe: S => Unit = (_:S) => ()): Observable[T] =
```

Comment:
Ah thanks! Didn't know/think of that semantic.

Errors should generally less confusing, as generally the state won't be of type `Unit` so there's no automatic conversion from `() => Unit` to `Unit` as in the issue you linked. The stateless variants don't have a generator.

An advantage of `() => T` is that it makes it very explicit that it is a function (thus it will be executed, possible multiple times). I would not expect `bar()` to be run multiple times in `foo(bar())`, even if the signature is `foo(x: => T)`, knowledge of the inner working of `foo` is necessary to know that.

I'll let you/the maintainers decide.